sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/AmqpMetricsProvider.java

```diff
@@ -0,0 +1,430 @@
+        }
+    }
+
+    /**
+     * Records the message was received.
+     */
+    public void recordReceivedMessage(Message message) {
+        if (!isEnabled || !receivedLag.isEnabled()
+            || message == null
+            || message.getMessageAnnotations() == null
+            || message.getBody() == null) {
+            return;
+        }
+
+        Map<Symbol, Object> properties = message.getMessageAnnotations().getValue();
+        Object enqueuedTimeDate = properties != null ? properties.get(ENQUEUED_TIME_ANNOTATION) : null;
+        if (enqueuedTimeDate instanceof Date) {
+            Instant enqueuedTime = ((Date) enqueuedTimeDate).toInstant();
+            long deltaMs = Instant.now().toEpochMilli() - enqueuedTime.toEpochMilli();
+            if (deltaMs < 0) {
+                deltaMs = 0;
+            }
+            receivedLag.record(deltaMs / 1000d, commonAttributes, Context.NONE);
+        } else if (enqueuedTimeDate != null) {
+            LOGGER.verbose("Received message has unexpected `x-opt-enqueued-time` annotation value - `{}`. Ignoring it.", enqueuedTimeDate);
+        }
+    }
+
+    /**
+     * Records that credits were added to link
+     */
+    public void recordAddCredits(int credits) {
+        if (isEnabled && addCredits.isEnabled()) {
+            addCredits.add(credits, commonAttributes, Context.NONE);
+        }
+    }
+
+    /**
+     * Records link error. Noop if condition is null (no error).
+     */
+    public void recordHandlerError(ErrorSource source, ErrorCondition condition) {
+        if (isEnabled && condition != null && condition.getCondition() != null) {
+            TelemetryAttributes attributes = amqpErrorAttributeCache.getOrCreate(condition.getCondition().toString());
+            switch (source) {
+                case LINK:
+                    if (linkErrors.isEnabled()) {
+                        linkErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                case SESSION:
+                    if (sessionErrors.isEnabled()) {
+                        sessionErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                case TRANSPORT:
+                    if (transportErrors.isEnabled()) {
+                        transportErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                default:
+                    LOGGER.verbose("Unexpected error source: {}", source);
+            }
+        }
+    }
+    private TelemetryAttributes getDeliveryStateAttribute(DeliveryState.DeliveryStateType state) {
+        // if there was no response, state is null and indicates a network (probably) error.
+        // we don't have an enum for network issues and metric attributes cannot have arbitrary
+        // high-cardinality data, so we'll just use vague "error" for it.
+        int ind = state == null ? DELIVERY_STATES_COUNT - 1 : state.ordinal();
```

Comment:
I am curious how this will version when we add more values to `DeliveryState.DeliveryStateType`. Will we then need to correlate the library version to the ordinal? `DELIVERY_STATES_COUNT - 1` could be the vague "error" state in v1, but when we add a new enum value, that would be a valid enum in v2, v1's values may look like valid response states.